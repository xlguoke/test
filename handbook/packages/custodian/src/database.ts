import { LoggerWrapper } from '@ilis/cordova-plugin-log4c'

export type BatchStatement = [string, any[]][]

/**
 * A batch execution transaction wrapper object
 */
export interface BatchTransaction {
  /**
   * sql statement and values
   * @param sql
   * @param params
   */
  execute(sql: string, params?: unknown[]): void

  /**
   * Promise-based commit method
   *
   * Must call this method after set statements
   */
  commit(): Promise<void>
}

/**
 * redundant open database cache
 *
 * The SQLite plugin already implements its own caching mechanism for opened databases.
 * This custom cache is primarily used to reduce debug messages.
 */
const openedDBs: Record<string, SQLitePlugin.Database> = {}
const log = new LoggerWrapper('custodian.db')

export function useDatabase(name?: string) {
  if (!name) {
    const userDbName = retrieveUserDatabase()
    if (!userDbName)
      throw new Error('you didn\'t provide the database name, and can\'t find user db name in the localstorage by contract')
    name = userDbName
  }
  if (name in openedDBs) {
    return openedDBs[name]
  }
  else {
    const db = sqlitePlugin.openDatabase(
      { name, location: 'default' },
      () => {},
      (err) => {
        log.error('failed to open db, {}', err)
        throw err
      },
    )
    openedDBs[name] = db
    return db
  }
}

/**
 * user store definition
 * @example
 * {
 *   username: 'foo'
 *   database: 'company.foo.db'
 * }
 */
function retrieveUserDatabase(): string | null {
  const users = localStorage.getItem('uhb.user')
  if (!users)
    return null
  const user = JSON.parse(users)
  if (!user)
    return null
  return user.database
}

/**
 * runtime type-safe should be considered
 * @param rs
 * @param mapper a mapper function to customize object transforming
 */
export function mapResultSet<T>(rs: SQLitePlugin.Results, mapper?: (t: any) => T) {
  const results: T[] = []
  for (let i = 0; i < rs.rows.length; i++)
    results.push(mapper ? mapper(rs.rows.item(i)) : rs.rows.item(i))
  return results
}

export async function executeSql(sql: string, params?: any[]): Promise<SQLitePlugin.Results> {
  const db = useDatabase()
  return new Promise((resolve, reject) => {
    db.executeSql(sql, params, rs => resolve(rs), reject)
  })
}

export async function queryForList<T>(sql: string, params?: any[], mapper?: (t: any) => T): Promise<T[]> {
  const rs = await executeSql(sql, params)
  return mapResultSet<T>(rs, mapper)
}

/**
 * find exactly one
 * @param sql
 * @param params
 * @throws Error if result size is not 1
 */
export async function queryForObject<T>(sql: string, params?: any[]) {
  const result = await queryForList<T>(sql, params)
  if (!result.length || result.length > 1)
    throw new Error(`Incorrect result size: expected 1, actual ${result.length}`)
  else
    return result[0]
}

export async function safeQueryForObject<T>(sql: string, params?: any[]) {
  try {
    return await queryForObject<T>(sql, params)
  }
  catch (e: any) {
    log.warn('Query for object encountered error: {}', 'message' in e ? e.message : e)
  }
}

export async function executeBatchSql(statements: BatchStatement, dbName?: string): Promise<void> {
  const db = useDatabase(dbName)
  return new Promise((resolve, reject) => {
    db.sqlBatch(statements, resolve, reject)
  })
}

export function newBatchTransaction(dbName?: string): BatchTransaction {
  const statements: BatchStatement = []
  return {
    execute(sql: string, params?: unknown[]): void {
      statements.push([sql, params || []])
    },
    commit(): Promise<void> {
      if (!statements.length)
        throw new Error('invalid state, got empty statements')
      return executeBatchSql(statements, dbName)
    },
  }
}
