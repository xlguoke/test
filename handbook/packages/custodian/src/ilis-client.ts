import { MD5 } from 'crypto-es/lib/md5'
import { z } from 'zod'

const client = z.object({
  /**
   * the name of the client
   */
  name: z.string(),
  /**
   * company code
   */
  code: z.string(),
  /**
   * server information
   */
  server: z.object({
    name: z.string(),
    addr: z.string().url('got an invalid ilis instance server url'),
  }),
})

type Client = z.infer<typeof client>

class ClientQuery {
  private readonly base: string

  constructor(base: string) {
    this.base = base
  }

  /**
   * make an ilis client request
   * @param code ilis company code
   * @return {Client}
   */
  async query(code: string): Promise<Client> {
    const url = new URL(`/api/instances/${code}`, this.base)
    const nonce = Date.now()
    const response = await fetch(url, {
      headers: {
        'X-Auth': 'ilis',
        'X-Auth-Nonce': `${nonce}`,
        'X-Auth-Sign': this.sign(nonce),
      },
    })
    if (response.ok)
      return client.parse(await response.json())
    throw new Error(await response.text())
  }

  private sign(nonce: number) {
    return MD5(`121f42460bc15b2666d84cc6f5de6ed1${nonce}`).toString()
  }
}

/**
 * Create an instance of ilis client query
 * @param auth base path of the ilis authorization center
 */
export function createClientQuery(auth: string): ClientQuery {
  return new ClientQuery(auth)
}
