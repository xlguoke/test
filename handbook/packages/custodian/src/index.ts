export * from './types'
export * from './ilis-client'
export * from './platform'
export * from './template'
export * from './database'
export * from './sample&param'
export * from './criterion'
export * from './contract&section'
export * from './project'
export * from './recordset'
export * from './miscellaneous'
export * from './task'
