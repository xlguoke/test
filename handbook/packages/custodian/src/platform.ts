/**
 * The version of the Cordova platform
 */
export type Version = string

/**
 * The ID of the Cordova platform
 */
export type PlatformID = string

/**
 * Get the version of the Cordova platform
 */
export function platformVersion(): Version {
  return cordova?.version || 'unknown'
}

/**
 * Get the ID of the Cordova platform which is the name of the operating system
 */
export function platformId(): PlatformID {
  return cordova?.platformId || 'unknown'
}
