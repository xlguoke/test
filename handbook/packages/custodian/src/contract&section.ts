import type { basic } from './types'
import { newBatchTransaction, queryForList } from './database'

export type Section = Omit<basic.Section, 'isDelete'>

/**
 * 保存合同段
 * - 删除被标记为删除的合同段
 * - 批量保存/更新合同段
 * @param contractParts
 */
export async function saveContractParts(contractParts: basic.ContractPart[]) {
  const tx = newBatchTransaction()
  for (const contractPart of contractParts) {
    if (contractPart.isDelete) {
      tx.execute('delete from contract_part where id = ?', [contractPart.id])
    }
    else {
      tx.execute(
        `insert into contract_part(id, contractPartName, projectId)
         VALUES (?, ?, ?)
         on conflict do update set contractPartName = ?,
                                   projectId = ?`,
        [
          contractPart.id,
          contractPart.contractPartName,
          contractPart.projectId,
          contractPart.contractPartName,
          contractPart.projectId,
        ],
      )
    }
  }
  await tx.commit()
}

/**
 * 保存工程划分
 * - 删除被标记为删除的工程划分
 * - 批量保存/更新工程划分
 * @param sections
 */
export async function saveSections(sections: basic.Section[]) {
  const tx = newBatchTransaction()
  for (const section of sections) {
    if (section.isDelete) {
      tx.execute('delete from section where id = ?', [section.id])
    }
    else {
      tx.execute(
        `insert into section(id, belongsId, unitProjectName, unitProjectType, levelCode, levelCodeLength)
         VALUES (?, ?, ?, ?, ?, ?)
         on conflict do update set belongsId = ?,
                                   unitProjectName = ?,
                                   unitProjectType = ?,
                                   levelCode = ?,
                                   levelCodeLength = ?`,
        [
          section.id,
          section.belongsId,
          section.unitProjectName,
          section.unitProjectType,
          section.levelCode,
          section.levelCodeLength,
          section.belongsId,
          section.unitProjectName,
          section.unitProjectType,
          section.levelCode,
          section.levelCodeLength,
        ],
      )
    }
  }
  await tx.commit()
}

/**
 * 根据工程项目获取工程划分，因ilis的工程划分既可以设置到工程项目亦可以设置到合同段
 * 目前手簿仅支持获取工程项目下的工程划分
 * @param projectId 工程项目ID
 */
export async function findSections(projectId: string) {
  return queryForList<Section>(
    `select section.* from section where belongsId = ?`,
    [projectId],
  )
}
