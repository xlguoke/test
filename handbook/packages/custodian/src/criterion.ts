import type { basic } from './types'
import { newBatchTransaction, queryForList, queryForObject } from './database'
import { Pageable } from './types'

export type Criterion = basic.Standard

/**
 * 先删除已下载的规程数据再执行批量插入
 * @param standards
 */
export async function saveCriteria(standards: basic.Standard[]) {
  const tx = newBatchTransaction()
  tx.execute('delete from standard where 1')
  for (const standard of standards) {
    tx.execute(
      'insert into standard(standardId, standardName, uniqueKey, publishCode, publishDate, expireDate, clauseCode) values (?, ?, ?, ?, ?, ?, ?) on conflict do nothing',
      [standard.standardId, standard.standardName, standard.uniqueKey, standard.publishCode, standard.publishDate, standard.expireDate, standard.clauseCode],
    )
  }
  await tx.commit()
}

export class CriterionQuery extends Pageable<Criterion> {
  constructor(public fuzzy: string | undefined = undefined) {
    super()
  }
}

/**
 * 分页查询本地规程库
 * @param query 模糊查询，支持规程名称、颁布号模糊查询
 */
export async function findCriteria(query?: CriterionQuery) {
  query = query || new CriterionQuery()
  if (query.fuzzy) {
    const fuzzy = `%${query.fuzzy.trim()}%`
    return queryForList<Criterion>(
      'select * from standard where standardName like ? or publishCode like ? limit ? offset ?',
      [fuzzy, fuzzy, query.limit, query.offset],
    )
  }
  return queryForList<Criterion>('select * from standard limit ? offset ?', [query.limit, query.offset])
}

export async function countCriteria(query?: CriterionQuery) {
  query = query || new CriterionQuery()
  let sql = 'select count(*) as count from standard'
  const params = []
  if (query.fuzzy) {
    const fuzzy = `%${query.fuzzy.trim()}%`
    sql += ' where standardName like ? or publishCode like ?'
    params.push(fuzzy, fuzzy)
  }
  return (await queryForObject<{ count: number }>(sql, params)).count
}
