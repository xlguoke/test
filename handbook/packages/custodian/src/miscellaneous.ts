import { z } from 'zod'
import { LoggerWrapper } from '@ilis/cordova-plugin-log4c'
import type { basic } from './types'
import { executeSql, queryForObject, safeQueryForObject } from './database'
import { useDir } from './template'

const log = new LoggerWrapper('custodian.miscellaneous')
/**
 * trim the `/UDR` suffix if present
 * @param url
 */
export function normalizeHitekCenterUrl(url: string) {
  return new URL('/', url).href
}

/**
 * 保存数据中心地址
 * @param url
 */
export async function saveHitekCenterUrl(url: string) {
  try {
    const existed = await getHitekCenterUrl()
    if (existed !== url)
      await executeSql('update conf set hitekCenter = ? where 1', [url])
  }
  catch (e) {
    await executeSql('insert into conf values (?)', [url])
  }
}

/**
 * 获取数据中心地址
 */
export async function getHitekCenterUrl() {
  const conf = await queryForObject<{ hitekCenter: string }>(
    'select hitekCenter from conf',
  )
  return conf.hitekCenter
}

/**
 * 保存数据版本信息
 * @param dataVersion
 */
export async function saveDataVersion(dataVersion: basic.DataVersion) {
  await executeSql(
    'insert into data_version values (?, ?, ?) on conflict do update set version = ?, paramVersionId = ?',
    [dataVersion.type, dataVersion.version, dataVersion.paramVersionId, dataVersion.version, dataVersion.paramVersionId],
  )
}

/**
 * 获取类型当前版本信息
 * @param type
 */
export async function getDataVersion(type: basic.DataVersion['type']) {
  return safeQueryForObject<basic.DataVersion>('select * from data_version where type = ?', [type])
}

const commonFileResponse = z.object({
  success: z.boolean(),
  msg: z.string().nullable(),
  data: z.array(z.object({
    fileName: z.string(),
    filePath: z.string(),
    lastWriteTime: z.string(),
    createTime: z.string(),
  })),
})

interface HitekCenter {
  api: string
  udr: string
}

export async function downloadExtraFilesForUdr(server: HitekCenter) {
  const url = new URL('/api/TestTemplate/GetAppTemplateCommonFiles', server.api)
  const res = await fetch(url)
  if (!res.ok)
    throw new Error(`failed to fetch app template common files from ${url.href}`)
  const result = commonFileResponse.parse(await res.json())
  const files: [string, string][] = []
  for (const datum of result.data) {
    const path = new URL(datum.filePath, server.udr)
    const fragments = decodeURI(path.pathname).split('/')
    const filename = fragments.pop()
    if (filename !== datum.fileName)
      throw new Error(`${filename}, ${datum.fileName} mismatch`)
    let dir: DirectoryEntry | undefined
    for (const fragment of fragments)
      dir = await useDir(fragment, dir)
    const fileEntry = await getFileEntry(dir!, filename)
    const file = await getFile(fileEntry)
    const lastModified = new Date(datum.lastWriteTime)
    if (!file.size || lastModified.getTime() < file.lastModified) {
      log.info(`download {} from {}`, filename, path)
      fetch(path)
        .then(res => res.blob())
        .then(res => fileEntry.createWriter(writer => writer.write(res)))
      files.push([filename, fileEntry.fullPath])
    }
  }
  return files
}

async function getFileEntry(dir: DirectoryEntry, file: string): Promise<FileEntry> {
  return new Promise((resolve, reject) => {
    dir.getFile(file, { create: true }, resolve, reject)
  })
}

async function getFile(entry: FileEntry): Promise<File> {
  return new Promise((resolve, reject) => {
    entry.file(resolve, reject)
  })
}

/**
 * Because the pre-defined variable {cordova.file.dataDirectory} cannot be used in the browser,
 * therefore, you'll get a rejection when running this code in the browser.
 */
export async function readLogEntries(): Promise<FileEntry[]> {
  return new Promise((resolve, reject) => {
    window.resolveLocalFileSystemURL(
      cordova.file.dataDirectory,
      (entry) => {
        if (entry.isFile)
          reject(new Error(`you'll never see me`))
        const dirReader = (entry as DirectoryEntry).createReader()
        dirReader.readEntries((entries) => {
          const logs = entries
            .filter(entry => entry.isFile && entry.name.endsWith('.log'))
            .map(entry => entry as FileEntry)
          resolve(logs)
        }, reject)
      },
      reject,
    )
  })
}

export function readLog(name: string): Promise<string> {
  return new Promise((resolve, reject) => {
    window.resolveLocalFileSystemURL(
      `${cordova.file.dataDirectory}/${name}`,
      (entry) => {
        if (entry.isDirectory)
          reject(new Error('wrong entry type, expect file'))
        const log = entry as FileEntry
        log.file((file) => {
          const fileReader = new FileReader()
          fileReader.onload = () => {
            resolve(fileReader.result as string)
          }
          fileReader.onerror = reject
          fileReader.readAsText(file)
        })
      },
      reject,
    )
  })
}
