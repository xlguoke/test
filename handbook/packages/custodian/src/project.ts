import type { basic } from './types'
import { Pageable } from './types'
import { newBatchTransaction, queryForList, queryForObject } from './database'

export type Project = Omit<basic.Project, 'isDeleted'>

/**
 * 保存工程项目
 * - 删除被标记为删除的工程项目
 * - 保存/更新工程项目
 * @param projects
 */
export async function saveProjects(projects: basic.Project[]) {
  const tx = newBatchTransaction()
  for (const project of projects) {
    if (project.isDeleted) {
      tx.execute('delete from project where id = ?', [project.id])
    }
    else {
      tx.execute(
        `insert into project(id, name, createDate, projectCode, constructionUnitName, buildUnitName, buildProjectName,
                             witnessUnitName, description, isCompleted, departNames)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
         on conflict do update set name                 = ?,
                                   projectCode          = ?,
                                   constructionUnitName = ?,
                                   buildUnitName        = ?,
                                   buildProjectName     = ?,
                                   witnessUnitName      = ?,
                                   description          = ?,
                                   isCompleted          = ?,
                                   departNames          = ?`,
        [
          project.id,
          project.name,
          project.createDate,
          project.projectCode,
          project.constructionUnitName,
          project.buildUnitName,
          project.buildProjectName,
          project.witnessUnitName,
          project.description,
          project.isCompleted,
          project.departNames,
          project.name,
          project.projectCode,
          project.constructionUnitName,
          project.buildUnitName,
          project.buildProjectName,
          project.witnessUnitName,
          project.description,
          project.isCompleted,
          project.departNames,
        ],
      )
    }
  }
  await tx.commit()
}

export class ProjectQuery extends Pageable<Project> {
  fuzzy: string | undefined
  constructor(fuzzy: string | undefined = undefined) {
    super()
    this.fuzzy = fuzzy
  }
}

/**
 * 分页查询工程项目
 * @param query 模糊查询，支持工程项目名称/项目编号
 */
export async function findProjects(query?: ProjectQuery) {
  query = query || new ProjectQuery()
  if (query.fuzzy) {
    const fuzzy = `%${query.fuzzy.trim()}%`
    return queryForList<Project>(
      `select * from project where name like ? or projectCode like ? limit ? offset ?`,
      [fuzzy, fuzzy, query.limit, query.offset],
    )
  }
  return queryForList<Project>(`select * from project limit ? offset ?`, [query.limit, query.offset])
}

export async function countProjects(query?: ProjectQuery) {
  query = query || new ProjectQuery()
  let sql = 'select count(*) as count from project'
  const params = []
  if (query.fuzzy) {
    const fuzzy = `%${query.fuzzy.trim()}%`
    sql += ' where name like ? or projectCode like ?'
    params.push(fuzzy, fuzzy)
  }
  return (await queryForObject<{ count: number }>(sql, params)).count
}
