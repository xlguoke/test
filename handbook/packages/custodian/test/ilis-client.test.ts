import { describe, expect, it } from 'vitest'
import { setupMockServerWithHandlers } from '../../../mocks/server'
import handlers from '../../../mocks/admin'
import { createClientQuery } from '../src'

const clientQuery = createClientQuery('https://auth.ilis.cn')

setupMockServerWithHandlers(handlers)

describe('ilis client query', () => {
  it('throws error when query a nonexistent instance', async () => {
    await expect(clientQuery.query('none exist')).rejects.toThrow('Client Not Found')
  })
  it('throws error when query an invalid instance', async () => {
    await expect(clientQuery.query('bar')).rejects.toMatchSnapshot()
  })
  it('returns the correct ilis client brevity information', async () => {
    const result = await clientQuery.query('foo')
    expect(result).toEqual({
      name: 'A valid instance',
      code: 'foo',
      server: {
        name: 'valid instance',
        addr: 'https://pro.ilis.cn',
      },
    })
  })
})
