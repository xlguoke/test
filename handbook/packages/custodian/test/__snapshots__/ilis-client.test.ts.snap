// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ilis client query > throws error when query an invalid instance 1`] = `
[ZodError: [
  {
    "validation": "url",
    "code": "invalid_string",
    "message": "got an invalid ilis instance server url",
    "path": [
      "server",
      "addr"
    ]
  }
]]
`;
